File: crates/clang-importer-sys/rust/src/ffi/swift/diagnostic_engine.rs
#[cxx::bridge(namespace = "rust::swift")]
pub(crate) mod ffi {
    struct DiagnosticEngine {
        ptr: UniquePtr<CxxDiagnosticEngine>,
    }

    #[namespace = "swift"]
    extern "C++" {
        include!("swift/AST/DiagnosticEngine.h");

        #[cxx_name = "DiagnosticEngine"]
        type CxxDiagnosticEngine;
    }

    #[namespace = "cxx::swift::DiagnosticEngine"]
    extern "C++" {
        include!("cxx/swift/DiagnosticEngine.hxx");

        #[namespace = "rust::swift"]
        type SourceManager = crate::swift::SourceManager;

        unsafe fn make(source_mgr: &mut SourceManager) -> UniquePtr<CxxDiagnosticEngine>;
    }
}

use self::ffi::{CxxDiagnosticEngine, DiagnosticEngine};
use crate::swift::SourceManager;
use cxx::UniquePtr;

impl From<UniquePtr<CxxDiagnosticEngine>> for DiagnosticEngine {
    #[inline]
    fn from(ptr: UniquePtr<CxxDiagnosticEngine>) -> Self {
        Self { ptr }
    }
}

impl DiagnosticEngine {
    pub unsafe fn new(source_mgr: &mut SourceManager) -> Self {
        let ptr = self::ffi::make(source_mgr);
        Self { ptr }
    }
}

File: crates/clang-importer-sys/cxx/lib/cxx/swift/DiagnosticEngine.cxx
#include "cxx/swift/DiagnosticEngine.hxx"

namespace cxx {
namespace swift {
namespace DiagnosticEngine {

std::unique_ptr<::swift::DiagnosticEngine>
make(rust::swift::SourceManager &SourceMgr) {
  ::swift::SourceManager &mgr = *SourceMgr.ptr;
  ::swift::DiagnosticEngine *diag = new ::swift::DiagnosticEngine(mgr);
  return std::unique_ptr<::swift::DiagnosticEngine>(diag);
}

} // namespace DiagnosticEngine
} // namespace swift
} // namespace cxx